kymatio/scattering2d/backend/numpy_backend.py

```diff
@@ -0,0 +1,176 @@
+        self.pre_pad = pre_pad
+        self.pad_size = pad_size
+        self.build()
+
+    def build(self):
+        self.np_pad = ((self.pad_size[0], self.pad_size[1]), (self.pad_size[2], self.pad_size[3]))
+
+    def __call__(self, x):
+        if self.pre_pad:
+            return x
+        else:
+            return np.pad(x, ((0,0), (0,0), self.np_pad[0], self.np_pad[1]), mode='reflect')
+
+def unpad(in_):
+    """
+        Slices the input tensor at indices between 1::-1
+        Parameters
+        ----------
+        in_ : tensor_like
+            input tensor
+        Returns
+        -------
+        in_[..., 1:-1, 1:-1]
+    """
+    return np.expand_dims(in_[..., 1:-1, 1:-1],-3)
+
+class SubsampleFourier(object):
+    """
+        Subsampling of a 2D image performed in the Fourier domain
+        Subsampling in the spatial domain amounts to periodization
+        in the Fourier domain, hence the formula.
+        Parameters
+        ----------
+        x : tensor_like
+            input tensor with at least 5 dimensions, the last being the real
+             and imaginary parts.
+            Ideally, the last dimension should be a power of 2 to avoid errors.
+        k : int
+            integer such that x is subsampled by 2**k along the spatial variables.
+        Returns
+        -------
+        res : tensor_like
+            tensor such that its fourier transform is the Fourier
+            transform of a subsampled version of x, i.e. in
+            FFT^{-1}(res)[u1, u2] = FFT^{-1}(x)[u1 * (2**k), u2 * (2**k)]
+    """
+    def __call__(self, x, k):
+        out = np.zeros((x.shape[0], x.shape[1], x.shape[2] // k, x.shape[3] // k), dtype=x.dtype)
+
+        y = x.reshape((x.shape[0], x.shape[1],
+                       x.shape[2]//out.shape[2], out.shape[2],
+                       x.shape[3]//out.shape[3], out.shape[3]))
+
+        out = y.mean(axis=(2, 4))
+        return out
+
+
+class Modulus(object):
+    """
+        This class implements a modulus transform for complex numbers.
+        Usage
+        -----
+        modulus = Modulus()
+        x_mod = modulus(x)
+        Parameters
+        ---------
+        x: input complex tensor.
+        Returns
+        -------
+        output: a tensor with imaginary part set to 0, real part set equal to
+        the modulus of x.
+    """
+    def __call__(self, x):
+        norm = np.abs(x)
+        return norm
+
+
+
+
+def fft(x, direction='C2C', inverse=False):
+    """
+        Interface with torch FFT routines for 2D signals.
+        Example
+        -------
+        x = torch.randn(128, 32, 32, 2)
+        x_fft = fft(x, inverse=True)
+        Parameters
+        ----------
+        input : tensor
+            complex input for the FFT
+        direction : string
+            'C2R' for complex to real, 'C2C' for complex to complex
+        inverse : bool
+            True for computing the inverse FFT.
+            NB : if direction is equal to 'C2R', then the transform
```

Comment:
then `inverse` should be `None` by default in the prototype. We shouldn't have a documented default value for a kwarg that we end up overriding in another kwarg.

Reply:
Seconded - there should be an error instead of changing something automatically

Reply:
I do not understand what you mean, can you provide me an example?

Reply:
Basically, what this docstring says is that _if `direction='C2R'`, I will ignore what you say in `inverse` and just set it to `True`._ What it should do instead is raise an error when `direction='C2R'` and `inverse=False`.
If you want to keep this somewhat sneaky behavior, then you can do what Vincent suggested and set `inverse=None` as default, which will be interpreted as `False` when `direction='C2C'`, and `True` when `direction='C2R'`.
I think I'd honestly just prefer an error to be raised and no `None` logic to be there

Reply:
As far as I can tell from the code, this is implemented exactly like that. So all that needs to change is the docstring, saying that `C2R` requires `inverse=True`

Reply:
OH ok.